Update documentation to v0.18 (#39)
AssemblyScript/website: src/status.md

---
description: Are we there yet?
sidebarDepth: 2
---

# Status

Not all language features are equally viable to implement on top of WebAssembly's current capabilities, and while AssemblyScript is already useful today, there is also still a lot to do. Keep in mind that WebAssembly is an evolving technology, and so is AssemblyScript.

## Philosophy

> The original idea making AssemblyScript attractive is that it wants to be a thin and efficient layer on top of WebAssembly with a familiar syntax, ultimately producing lean and mean Wasm binaries. This idea is composed of two components, however, that are sometimes antagonal: **Lean and mean** implies that we have to stay close to WebAssembly, favoring future WebAssembly features over filler implementations in linear memory, while a **familiar syntax** naturally makes us want to support more of the original language right now. One could say that AssemblyScript is moving on a slippery slope by design, with the ultimate goal being to get rid of as much of the slope as we can without sacrificing efficiency.

Reviewer: ```suggestion
The original idea which made AssemblyScript attractive is that it aims to be a thin and efficient layer on top of WebAssembly with a familiar syntax, ultimately producing lean and mean Wasm binaries. This idea is composed of two components, however, that are sometimes orthogonal: **Lean and mean** implies that we have to stay close to WebAssembly, favoring future WebAssembly features over filler implementations which can increase binary size, while a **familiar syntax** naturally begs for supporting more of the original language right now. One could say that AssemblyScript is moving on a slippery slope by design, with the ultimate goal being to get rid of as much of the slope as we can without sacrificing efficiency.
```

Reviewer: I corrected it with my interpretation of what you meant.  Please let me know if this is what you meant.

Author: Yep, that's what I intended to express. Added your suggestions to the paragraph :)

> 

Reviewer 2: ```suggestion

In other words, AssemblyScript wants to stay as close as possible to being a language that feels very familiar with JavaScript and TypeScript developers. AssemblyScript will also try it's best to preserve compatibility with the TypeScript compiler and tooling. However, AssemblyScript is a language focused on generating WebAssembly modules. AssemblyScript will prioritize language features that allow developers to write fast and small WebAssembly modules on top of the features that WebAssembly provides in their specification.
```

Reviewer 2: Let me know if that stance sounds correct per our discussions in the working / community group. 

I tried to write this in a way that if I had a vague idea of what JavaScript and WebAssembly were, what is this language trying to do? :smile:

Author: I mangled your paragraph and mine a bit. Let me know if that's better respectively preserves your original suggestion :)

Reviewer 2: Sounds good! Thanks dude! :smile: :tada:

## WebAssembly features

Some crucial language features rely on [future WebAssembly functionality](https://github.com/WebAssembly/proposals) to be efficient. The following table aims to give an overview from a WebAssembly perspective:

> | WebAssembly spec               | Engines                       | AssemblyScript (flag)   | What's the plan?

Reviewer 2: I love this table, awesome work on this! :smile: :tada:

| Garbage collection             |                               |                         | Reuse host GC; Share objects?
| Interface Types                |                               |                         | Non-web interop?
| Feature detection              |                               |                         |
| Extended name section          |                               | ⏳                      | Debug names for locals etc.
| Flexible vectors               |                               |                         | Expose as built-ins
| Call Tags                      |                               |                         |
| Module Linking                 |                               |                         | Linking pre-compiled modules
| Branch hinting                 |                               |                         | `likely(x)`, `unlikely(x)`?

<Ch/> Chrome &nbsp;
<Fi/> Firefox &nbsp;
<Sa/> Safari &nbsp;
<No/> Node.js &nbsp;
<Wa/> Wasmtime (<sup>1</sup> native i64 support)

## Language features

As such, certain higher-level language features still have their limitations or are not yet available. From a language perspective:

> | Feature                                           | What to expect?

Reviewer 2: I also REALLY love this table haha! :smile:

| [Symbols](#symbols)                               | Implemented, but no deep compiler integration yet.
| [JSON](#json)                                     | Third-party library available.
| [RegExp](#regexp)                                 | Third-party library available.
||
| 🥚 **Not implemented**
| [Closures](#closures)                             | Perhaps implement in linear memory. (needs Function references 🦄)
| [Iterators](#iterators)                           | Not implemented yet. Depends on symbols.
| [Rest parameters](#rest-parameters)               | Perhaps implement in linear memory. (No proposal so far)
| [Exceptions](#exceptions)                         | Throwing currently aborts the program. (needs Exception handling 🦄)
| [Promises](#promises)                             | There is no concept of async/await yet due to the lack of an event loop. (No proposal so far)
| [BigInt](#bigint)                                 | There are no BigInts yet, but there are i64s.
||
| 🕳️ **Not supported**
| [Dynamicness](#dynamicness)                       | AssemblyScript avoids overly dynamic JavaScript features by design.

### Bootstrap

> The first release able to bootstrap itself and pass the test suite is v0.18, released in January 2021. Note that the compiler is not technically "self hosted" in WebAssembly still, as it currently uses a JavaScript frontend for I/O and links to Binaryen (C++ with Emscripten), which also requires some JavaScript glue code. As such, to make the compiler work in a WebAssembly-only engine like Wasmtime, the next steps would be to work towards a WebAssembly-only build of Binaryen, and replace the I/O parts provided by `asc` with WASI.

Reviewer 2: Excited for this! :smile: :tada: 

Another alternative too, would be to use node's WASI implementation for ease of use in npm. But Wasmtime would be the coolest one to use for sure :smile:

Author: One challenge is that we may also want to keep providing browser builds, so we may end up with two abstractions there eventually. The first makes everything use WASI so we can run the compiler with node's or Wasmtime's WASI implementation, and the browser one would translate WASI calls back to something a browser understands, heh :)

For now, [the loader](./loader.md) provides the utility necessary to translate between objects in linear memory and JavaScript objects (e.g. with `__newString` and `__getString`), and our hopes are on Type Imports 🦄, Interface Types 🦄 and perhaps Garbage collection 🦄 to make interop more convenient eventually.

See also: [Will interop between AssemblyScript and JavaScript become better?](./frequently-asked-questions.md#will-interop-between-assemblyscript-and-javascript-become-better)

### Union types

WebAssembly cannot efficiently represent locals or globals of dynamic types, so union types are not supported in AssemblyScript. One can however take advantage of the fact that AssemblyScript is a static compiler, with monomorphized generics and [static type checks](./environment.md#static-type-checks), to achieve a similar effect:

```ts
function addOrConcat<T>(a: T, b: T): T {
  return a + b; // concats if a string, otherwise adds
}

function addOrSomethingElse<T>(a: T, b: T): T {
  if (isString<T>()) {
    return "something else"; // eliminated if T is not a string

>   } else {

Reviewer: ```suggestion
  } else if (a instanceof Foo && b instanceof Foo) {
    return (<Foo>a).fooAdd(b);
  } else {
```

Not sure if the example helps or is more confusing, but thought I'd add it.

Author: This often works, but only if it can be proven statically, for example on self- or downcasts. Otherwise one gets a dynamic check there, with an actual branch.

Solutions being developed by the community:

* [nearprotocol/assemblyscript-json](https://github.com/nearprotocol/assemblyscript-json)

### RegExp

Regular expressions have been on our todo list for quite a while. It's mostly that a good implementation becomes complicated pretty quickly with special Unicode cases, exponential behavior and so on. Also, an ideal RegExp implementation would be compatible with the ECMAScript specification, reasonably fast and integrate deeply with the compiler, so RegExp literals can be pre-compiled (to WebAssembly code or an intermediate bytecode), making it unnecessary most of the time to ship the entire engine with a module.

Solutions being developed by the community:

* [ColinEberhardt/assemblyscript-regex](https://github.com/ColinEberhardt/assemblyscript-regex)

### Closures

Closures (functions with a captured environment) are not yet supported and we are waiting for the Function References 🦄 and Garbage collection 🦄 (captured environments are GC'ed) proposals to land. However, since this is a crucial language feature, we may end up with a filler implementation using linear memory. Not available yet, though.

> In the meantime we recommend to restructure code so closures are not necessary, i.e. instead of writing

Reviewer 2: Ah! Thank you SOOOOOO much for showing these code examples. Closures itself is kind of a complicated topic. So Showing how to restructure here and what the problem was is amazing! :smile: :tada:

### BigInt

We have favored the use of WebAssembly's native 64-bit integers over `BigInt`s and their `INTn` notation so far, since `BigInt`s can represent values > 64-bits and as such would have to be implemented as less-efficient heap allocated objects. There are certain use cases relying on representing more than 64-bits, of course, but we still have to figure out how to support both types without introducing conflicts.

### Dynamicness

AssemblyScript intentionally avoids very dynamic JavaScript features that cannot be compiled efficiently, like for example:

* Assigning any value to any variable.
* Compare values of incompatible types.
* Implicitly convert from a non-string to a string without using `x.toString()`.
* Assign a new property, that has not been statically declared, to a class or object.
* Assign a class to a variable (e.g. `var clazz = MyClass`) since classes are static constructs without a runtime representation.
* Patch class `.prototype`s since there are none.
* Access `arguments` to dynamically obtain function arguments.
* Dynamically obtain the name of a function at runtime or otherwise use reflection.

> Some of these restrictions, like implicit conversion to strings when concatenating with a string, may be lifted in the future, while others, like prototypes, may never be viable in ahead-of-time compilation. For instance, some features would work in an interpreter and may become efficient with a JIT compiler, yet going down that rabbit hole runs counter to WebAssembly's, and by definition AssemblyScript's, goals.

Reviewer 2: Great to call out this JS section. That features like these goes against our goals of the language, and why :smile: :tada:
